README.md
<center>

# SWEEP

</center>
<p align="center">
    <img src="https://github.com/sweepai/sweep/assets/26889185/39d500fc-9276-402c-9ec7-3e61f57ad233">
</p>
<p align="center">
    <i>Bug Reports & Feature Requests ⟶&nbsp; Code Changes</i>
</p>
<a href="https://hub.docker.com/r/sweepai/sweep">
    <img alt="Docker Pulls" src="https://img.shields.io/docker/pulls/sweepai/sweep" />
</a>

<a href="https://docs.sweep.dev/">
    <img alt="Docs" src="https://img.shields.io/badge/Docs-docs.sweep.dev-blue?link=https%3A%2F%2Fdocs.sweep.dev">
</a>
<a href="https://discord.gg/sweep">
    <img src="https://dcbadge.vercel.app/api/server/sweep?style=flat" />
</a>
<a href="https://github.com/sweepai/sweep">
    <img src="https://img.shields.io/github/stars/sweepai/sweep" />
</a>
<a href="https://github.com/sweepai/sweep">
    <img src="https://img.shields.io/github/commit-activity/m/sweepai/sweep" />
</a>
<a href="https://twitter.com/sweep__ai">
    <img src="https://img.shields.io/twitter/url?url=https%3A%2F%2Ftwitter.com%2Fsweep__ai" />
</a>
</p>

*🎉 We recently changed our license to the Elastic License V2 to allow Sweep for commercial usage. Check out https://docs.sweep.dev/deployment*

---

<b>Sweep</b> is an AI junior developer that transforms bug reports & feature requests into code changes. :robot:

Describe bugs, small features, and refactors like you would to a junior developer, and Sweep:
1. Reads your codebase
2. Plans the changes
3. **Writes a pull request with code** ⚡

### Features

* Turns issues directly into pull requests (without an IDE)
* Addresses developer replies & comments on its PRs
* Uses embedding-based code & online document search
* Validates its changes with GitHub Actions and self-review

### How Sweep is different

Unlike Copilot, which only provides IDE-based autocompletion, Sweep handles the **entire flow end-to-end**. Unlike ChatGPT, Sweep is able to automatically understand and search through your code base, removing the need to tediously copy-and-paste files. Check out examples [here](https://docs.sweep.dev/about/examples)!

<details>
    <summary>
        Sweep vs. GPT-Engineer, Smol Developer and AutoGPT
    </summary>
    Sweep is built to improve on an existing codebase, which is a more frequent and higher need, than generating boilerplate, which is mostly a solved problem since you can just fork existing boilerplates.
</details>

<details>
    <summary>
        Sweep vs. Cody and Bloop
    </summary>
    We do more than just chat-with-your-code by actually writing code changes.
</details>

[Demo](https://github.com/sweepai/sweep/assets/44910023/365ec29f-7317-40a7-9b5e-0af02f2b0e47)

---

## Getting Started

Install Sweep by adding the [**Sweep GitHub App**](https://github.com/apps/sweep-ai) to your desired repositories.

* For more details, visit our [Installation](docs/installation.md) page.

* Note: Sweep only considers issues with the "Sweep:" title on creation and not on update. If you want Sweep to pick up an issue after it has been created, you can add the "Sweep" label to the issue.

* We support all languages GPT-4 supports, including Python, JS/TS, Rust, Go, Java, C# and C++.

## Self-hosting

You can self-host Sweep with the Docker image (`https://hub.docker.com/r/sweepai/sweep`). The setup instructions are at [Deployment](https://docs.sweep.dev/deployment).

## Limitations of Sweep

* **Large-scale refactors**: >3 files or >150 lines of code changes (we're working on this!)
    * e.g. Refactor the entire codebase from TensorFlow to PyTorch

* **Editing images** and other non-text assets
    * e.g. Use the logo to create favicons for our landing page

* **Performing actions involving a dashboard**, including fetching API tokens
    * e.g. Set up sign-in using Ethereum

## Development

### Starting the Webhook
1. Clone the repo with `git clone https://github.com/sweepai/sweep`.
2. Create `.env` according to https://docs.sweep.dev/deployment.
3. Run `docker compose up --build`. This will take a few moments to start.

To build the Docker images, run `docker compose build`.

---

## Story

We were frustrated by small tickets, like simple bug fixes, annoying refactors, and small features. Each task required us to open our IDE to fix simple bugs. So we decided to leverage the capabilities of ChatGPT to address this directly in GitHub.

Unlike existing AI solutions, this can solve entire tickets and can be parallelized + asynchronous: developers can spin up 10 tickets and Sweep will address them all at once.

## The Stack
- **GPT-4 32k** & GPT-3.5 16k
- GTE-base embedding model
- ActiveLoop DeepLake Vector DB
- Modal Labs for infra + deployment

## Highlights
Examine pull requests created by Sweep [here](https://docs.sweep.dev/about/examples).

## Pricing
Every user receives unlimited GPT-3.5 tickets and 5 GPT-4 tickets per month. To prevent abuse, users can use 2 GPT-4 tickets a day.

For professionals who want more tickets and priority support/feature requests, check out [Sweep Pro](https://buy.stripe.com/6oE5npbGVbhC97afZ4) now at $480/month. In addition, we're also offering 15 PRs for $60 as a one-time purchase for anyone interested in eventually purchasing Sweep Pro. You can purchase this [here](https://buy.stripe.com/7sI4jlaCR3PaabebIP).

## Roadmap
We plan on rapidly improving Sweep. To see what we're working on, check out our [Roadmap](https://docs.sweep.dev/roadmap).

---

## Star History

[![Star History Chart](https://api.star-history.com/svg?repos=sweepai/sweep&type=Date)](https://star-history.com/#sweepai/sweep&Date)

Consider starring us if you're using Sweep so more people hear about us!

## Contributing

Contributions are welcome and greatly appreciated! To get set up, clone the repo and run

```
poetry install
poetry shell
```

To build the docker image run

For detailed guidelines on how to contribute, please see the [CONTRIBUTING.md](CONTRIBUTING.md) file.
* [Sweep Docs](https://docs.sweep.dev/).


<h2 align="center">
    Contributors
</h2>
<p align="center">
    Thank you for your contribution!
</p>
<p align="center">
    <a href="https://github.com/sweepai/sweep/graphs/contributors">
      <img src="https://contrib.rocks/image?repo=sweepai/sweep" />
    </a>
</p>
<p align="center">
    and, of course, Sweep!
</p>


CHANGELOG.md
# Changelog

## August 8, 2023

- We just launched external documentation search! 📚 For example, if you want Sweep to use the OpenAI docs, you can mention “openai” in your issue and Sweep will search over the OpenAI docs for you. 👀

Right now we’ve manually indexed the documentation for frameworks like Langchain, ReactJS, and Docusaurus here: https://github.com/sweepai/sweep/blob/main/sweepai/core/documentation.py#L37-L47. You can mention those now to start using them!

We’ll also allow you to add your own documentation soon, so look out for that tomorrow 😀

The page contains code snippets from the file `sweepai/core/documentation.py`. The code includes imports from various modules and libraries such as `asyncio`, `re`, `deeplake.core.vectorstore.deeplake_vectorstore`, `loguru`, `tqdm`, `sweepai.core.lexical_search`, `sweepai.core.robots`, `sweepai.core.webscrape`, `sweepai.pre_indexed_docs`, and `sweepai.config.server`.

The code defines a class called `ModalEmbeddingFunction` with methods `__init__` and `__call__`. It also defines a class called `CPUEmbedding` with methods `__init__` and `compute`. Additionally, there is a function called `chunk_string`.

The code uses a constant `MODEL_DIR` and a batch size of 128. It also uses a timeout of 60 minutes. The `ModalEmbeddingFunction` class has a default batch size of 1024.

The `CPUEmbedding` class initializes a `SentenceTransformer` model and computes embeddings for a list of texts using the `encode` method. The `compute` method returns the computed embeddings.

The `chunk_string` function splits a string into sentences using regular expressions.

Overall, the code appears to be related to embedding functions and text processing.

## August 6, 2023

This was done by improving how Sweep chooses what to modify as well as the self-review process, where it would correct itself when it doesn't fully implement a class or function.

This was done by improving how Sweep chooses what to modify as well as the self-review process, where it would correct itself when it doesn't fully implement a class or function.

An example of this is at https://github.com/kevinlu1248/pyepsilla/pull/15 where Sweep had unimplemented changes but the self-review corrected itself.

The page is a pull request on GitHub titled "Create JavaScript/TypeScript client based on Python client" by sweep-canary[bot]. The pull request adds a JavaScript/TypeScript client for EpsillaDB, which is based on the functionality of the existing Python client. The new client includes a Client class in jsEpsilla/vectordb/client.ts that handles networking, makes requests, and processes responses. It also includes a Field class and a FieldType enum in jsEpsilla/vectordb/field.ts that mirror the functionality of the Python versions. The README.md file has been updated to include instructions on how to install and use the new JavaScript/TypeScript client, along with examples of how to use its methods. The pull request includes four commits and fixes issue #4.

## August 5, 2023

- William: ⏩ We just pushed fixes that should speed up Sweep's first step by a lot, especially for large repos (could have taken 5 minutes+, should now be sub 1 minute). Sweep was slow because of a hotfix we pushed due to our infra provider having issues. Now that Sweep is working reliably, we're working on getting the speed back 🚤
- William: ⏩ We just pushed fixes that should speed up Sweep's first step by a lot, especially for large repos (could have taken 5 minutes+, should now be sub 1 minute). Sweep was slow because of a hotfix we pushed due to our infra provider having issues. Now that Sweep is working reliably, we're working on getting the speed back 🚤
- William: Just pushed an improvement that should really improve latency. We were previously updating the cache for inactive users, which slows everything down. We now only do this for active and Sweep Pro users. :sweeping:
- Kevin:
- William: Hey everyone, we have some issues with modifying files (you'll see this error at 60%). Apologies, I'm on this now.
- William: ⏩ We just pushed fixes that should speed up Sweep's first step by a lot, especially for large repos (could have taken 5 minutes+, should now be sub 1 minute). Sweep was slow because of a hotfix we pushed due to our infra provider having issues. Now that Sweep is working reliably, we're working on getting the speed back 🚤
- Kevin: The erroneous GitHub Actions logs should be mostly fixed now. Let me know if you still get a GitHub Action with the wrong logs.

- Kevin: ⚡ We just released further optimization for search! We noticed that the queue times recently have been very long so for small updates on repos we use CPU to re-embed the new files, which is faster than the queue times!

## July 30, 2023

- We just launched external search! Sweep can now read the page of any link you reference in the issue title + description to publicly accessible URLs, including links to docs and PR's on open-source repos!
- We just launched external search! Sweep can now read the page of any link you reference in the issue title + description to publicly accessible URLs, including links to docs and PR's on open-source repos!

## July 29, 2023

- The search index now uses a higher quality retrieval embedding model!

src/slack.py
import json
import os

from fastapi import FastAPI, HTTPException, Request, Response
from fastapi.responses import RedirectResponse

import modal
from pydantic import BaseModel
from pymongo import MongoClient
import requests
from src.core.entities import FileChangeRequest, Function, PullRequest
import slack_sdk
from loguru import logger
from slack_sdk.oauth.installation_store import Installation, InstallationStore, Bot

from src.core.sweep_bot import SweepBot
from src.utils.github_utils import get_github_client
from src.utils.constants import API_NAME, BOT_TOKEN_NAME, PREFIX, SLACK_NAME
from src.core.prompts import slack_slash_command_prompt
from src.utils.github_utils import get_installation_id
import src.utils.event_logger

image = (
    modal.Image.debian_slim()
    .apt_install("git")
    .pip_install(
        "slack-sdk",
        "slack-bolt",
        "pymongo",
        "PyGithub",
        "gitpython",
        "openai",
        "anthropic",
        "loguru",
        "tqdm",
        "highlight-io",
        "posthog",
        "pyyaml",
    )
)
stub = modal.Stub(SLACK_NAME)
secrets = [
    modal.Secret.from_name("slack"),
    modal.Secret.from_name("openai-secret"),
    modal.Secret.from_name(BOT_TOKEN_NAME),
    modal.Secret.from_name("mongodb"),
]

functions = [
    Function(
        name="create_pr",
        description="Creates a PR.",
        parameters={
            "properties": {
                "plan": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "file_path": {
                                "type": "string",
                                "description": "The file path to change.",
                            },
                            "instructions": {
                                "type": "string",
                                "description": "Detailed description of what the change as a list.",
                            },
                        },
                        "required": ["file_path", "instructions"],
                    },
                    "description": "A list of files to modify or create and instructions for what to modify or create.",
                },
                "title": {
                    "type": "string",
                    "description": "Title of PR",
                },
                "summary": {
                    "type": "string",
                    "description": "Detailed summary of PR",
                },
                "branch": {
                    "type": "string",
                    "description": "Name of branch to create PR in.",
                },
            }
        },
    ),
    Function(
        name="get_relevant_snippets",
        description="Search engine for relevant snippets in the current repo in natural language.",
        parameters={
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Natural language query to search.",
                }
            }
        },
    ),
]

pr_format = """I'm going to create a PR with the following:

> *{title}*
> {summary}

I have the following plan:
{plan}

:hourglass_flowing_sand: Creating...
"""

pr_done_format = """:white_check_mark: Done creating PR at {url} with the following:
> *{title}*
> {summary}
"""

thread_query_format = "Message thread: {messages}"

# TODO(sweep): move to constants
slack_app_page = "https://sweepusers.slack.com/apps/A05D69L28HX-sweep"
slack_install_link = "https://slack.com/oauth/v2/authorize?client_id=5364586338420.5448326076609&scope=channels:read,chat:write,chat:write.public,commands,groups:read,im:read,incoming-webhook,mpim:read&user_scope="


class MongoDBInstallationStore(InstallationStore):
    def __init__(self):
        self.client = MongoClient(os.environ["MONGODB_URI"])
        self.db = self.client["slack"]
        self.installation_collection = self.db["installation"]
        self.bot_collection = self.db["bot"]

    def save(self, installation: Installation):
        self.installation_collection.replace_one(
            {
                "user_id": installation.user_id,
                "team_id": installation.team_id,
                "enterprise_id": installation.enterprise_id,
                "is_enterprise_install": installation.is_enterprise_install,
            },
            {
                "installation": installation.to_dict(),
                "user_id": installation.user_id,
                "team_id": installation.team_id,
                "enterprise_id": installation.enterprise_id,
                "is_enterprise_install": installation.is_enterprise_install,
                "prefix": PREFIX,
                "access_token": installation.bot_token,
            },
            upsert=True,
        )

    def save_bot(self, bot: Bot):
        self.bot_collection.replace_one(
            {
                "team_id": bot.team_id,
                "enterprise_id": bot.enterprise_id,
                "is_enterprise_install": bot.is_enterprise_install,
            },
            {
                "installation": bot.to_dict(),
                "team_id": bot.team_id,
                "enterprise_id": bot.enterprise_id,
                "is_enterprise_install": bot.is_enterprise_install,
                "prefix": PREFIX,
                "access_token": bot.bot_token,
            },
            upsert=True,
        )

    def find_installation(
        self,
        *,
        enterprise_id: str,
        team_id: str,
        is_enterprise_install: bool,
        user_id: str | None = None,
    ):
        if user_id:
            return Installation(
                **self.installation_collection.find_one(
                    {
                        "user_id": user_id,
                        "team_id": team_id,
                        "enterprise_id": enterprise_id,
                        "is_enterprise_install": is_enterprise_install,
                    }
                )["installation"]
            )
        else:
            return Installation(
                **self.installation_collection.find_one(
                    {
                        "team_id": team_id,
                        "enterprise_id": enterprise_id,
                        "is_enterprise_install": is_enterprise_install,
                    }
                )["installation"]
            )

    def find_bot(self, *enterprise_id: str, team_id: str, is_enterprise_install: bool):
        return Installation(
            **self.bot_collection.find_one(
                {
                    "enterprise_id": enterprise_id,
                    "team_id": team_id,
                    "is_enterprise_install": is_enterprise_install,
                }
            )["installation"]
        )

    def delete_installation(self, *, enterprise_id: str, team_id: str):
        self.installation_collection.delete_one(
            {
                "enterprise_id": enterprise_id,
                "team_id": team_id,
            }
        )

    def delete_bot(self, *, enterprise_id: str, team_id: str, user_id: str):
        self.installation_collection.delete_one(
            {"enterprise_id": enterprise_id, "team_id": team_id, "user_id": user_id}
        )


class SlackSlashCommandRequest(BaseModel):
    channel_name: str
    channel_id: str
    text: str
    user_name: str
    user_id: str
    team_id: str
    is_enterprise_install: bool
    enterprise_id: str | None = None


@stub.function(
    image=image,
    secrets=secrets,
    timeout=15 * 60,
)
def reply_slack(request: SlackSlashCommandRequest, thread_ts: str | None = None):
    try:
        create_pr = modal.Function.lookup(API_NAME, "create_pr")
        client = None
        installation_store = MongoDBInstallationStore()
        token = installation_store.find_installation(
            team_id=request.team_id,
            enterprise_id=request.enterprise_id,
            user_id=request.user_id,
            is_enterprise_install=request.is_enterprise_install,
        ).bot_token
        client = slack_sdk.WebClient(token=token)
    except Exception as e:
        logger.error(f"Error initializing Slack client: {e}")
        raise e

    try:
        channel_info = client.conversations_info(channel=request.channel_id)
        channel_description: str = channel_info["channel"]["purpose"]["value"]
        logger.info(f"Channel description: {channel_description}")

        repo_full_name = channel_description.split()[-1]
        logger.info(f"Repo name: {repo_full_name}")

        organization_name, repo_name = repo_full_name.split("/")

        try:
            installation_id = get_installation_id(organization_name)
            g = get_github_client(installation_id)
            repo = g.get_repo(repo_full_name)
        except Exception as e:
            # TODO: provide better instructions for installation
            client.chat_postMessage(
                channel=request.channel_id,
                text=f"An error has occurred with fetching the credentials for {repo_full_name}. Please ensure that the app is installed on the Github repo.",
            )
            raise e

        sweep_bot = SweepBot(repo=repo)

        search_already_done = bool(thread_ts)
        if not thread_ts:
            thread = client.chat_postMessage(
                channel=request.channel_id,
                text=f">{request.text}\n- <@{request.user_name}>",
            )
            thread_ts = thread["ts"]
            query = request.text
        else:
            messages = client.conversations_replies(
                channel=request.channel_id,
                ts=thread_ts,
            )["messages"]
            query = thread_query_format.format(
                messages="\n".join(
                    [
                        f"<message user={message['user']}>\n{message['text']}\n</message>"
                        for message in messages
                    ]
                )
            )
    except Exception as e:
        client.chat_postMessage(
            channel=request.channel_id,
            text=":exclamation: Sorry, something went wrong. Sometimes this is because the Github app is not installed on your repo.",
        )
        raise e

    try:
        if not search_already_done:
            logger.info("Fetching relevant snippets...")
            searching_message = client.chat_postMessage(
                channel=request.channel_id,
                text=":mag_right: Searching for relevant snippets...",
                thread_ts=thread_ts,
            )
            snippets = sweep_bot.search_snippets(
                request.text, installation_id=installation_id
            )
            message = ":mag_right: Some relevant snippets I found:\n\n"
            message += "\n".join(
                f"{snippet.get_slack_link(repo_name)}\n```{snippet.get_preview()}```"
                for snippet in snippets
            )
            client.chat_update(
                channel=request.channel_id,
                ts=searching_message["ts"],
                text=message,
            )
        else:
            snippets = []
        prompt = slack_slash_command_prompt.format(
            relevant_snippets="\n".join([snippet.xml for snippet in snippets]),
            relevant_directories="\n".join([snippet.file_path for snippet in snippets]),
            repo_name=repo_name,
            repo_description=repo.description,
            username=request.user_name,
            query=query,
        )
        response = sweep_bot.chat(
            prompt, functions=functions, function_name={"name": "create_pr"}
        )
        logger.info(response)

        while sweep_bot.messages[-1].function_call is not None:
            obj = sweep_bot.messages[-1].function_call
            name = obj["name"]
            arguments = json.loads(obj["arguments"])
            if name == "get_relevant_snippets":
                logger.info("Searching for relevant snippets...")
                search_message = client.chat_postMessage(
                    channel=request.channel_id,
                    text=f":mag_right: Searching \"{arguments['query']}\" in the codebase...",
                    thread_ts=thread_ts,
                )
                additional_snippets = sweep_bot.search_snippets(
                    arguments["query"], installation_id=installation_id
                )
                # additional_snippets = default_snippets
                additional_snippets_message = (
                    f":mag_right: Found {len(additional_snippets)} additional snippets with the query \"{arguments['query']}\":\n\n"
                    + "\n".join(
                        f"{snippet.get_slack_link(repo_name)}\n```{snippet.get_preview()}```"
                        for snippet in additional_snippets
                    )
                )
                client.chat_update(
                    channel=request.channel_id,
                    text=additional_snippets_message,
                    ts=search_message["ts"],
                )
                response = sweep_bot.chat(
                    additional_snippets_message, functions=functions
                )
            elif name == "create_pr":
                title = arguments["title"]
                summary = arguments["summary"]
                branch = arguments["branch"]
                plan = arguments["plan"]
                plan_message = "\n".join(
                    f"• `{file['file_path']}`: {file['instructions']}" for file in plan
                )
                plan_message = ">" + plan_message.replace("\n", "\n> ")

                creating_pr_message = client.chat_postMessage(
                    channel=request.channel_id,
                    text=pr_format.format(
                        title=title, summary=summary, plan=plan_message
                    ),
                    thread_ts=thread_ts,
                )
                file_change_requests = []
                for file in plan:
                    change_type = "create"
                    try:
                        contents = repo.get_contents(file["file_path"])
                        if contents:
                            change_type = "modify"
                    except:
                        pass
                    file_change_requests.append(
                        FileChangeRequest(
                            filename=file["file_path"],
                            instructions=file["instructions"],
                            change_type=change_type,
                        )
                    )
                pull_request = PullRequest(
                    title=title,
                    branch_name=branch,
                    content=summary,
                )
                results = create_pr.call(
                    file_change_requests=file_change_requests,
                    pull_request=pull_request,
                    sweep_bot=sweep_bot,
                    username=request.user_name,
                    installation_id=installation_id,
                )
                logger.debug(results)
                pr = results["pull_request"]
                client.chat_update(
                    channel=request.channel_id,
                    text=pr_done_format.format(
                        url=pr.html_url,
                        title=title,
                        summary=summary,
                    ),
                    ts=creating_pr_message["ts"],
                )
                break
            else:
                raise Exception(f"Unknown function {name}")
        else:
            # no breaks were called
            client.chat_postMessage(
                channel=request.channel_id, text=response, thread_ts=thread_ts
            )
    except Exception as e:
        client.chat_postMessage(
            channel=request.channel_id,
            text=":exclamation: Sorry, something went wrong.",
            thread_ts=thread_ts,
        )
        logger.error(f"Error creating PR: {e}")
        raise e


@stub.function(image=image, keep_warm=1)
@modal.web_endpoint(method="POST")
async def entrypoint(request: Request):
    body = await request.form()
    request = SlackSlashCommandRequest(**dict(body))
    reply_slack.spawn(request)
    return Response(status_code=200)


@stub.function(image=image, keep_warm=1)
@modal.web_endpoint(method="GET")
def install():
    return RedirectResponse(url=slack_install_link)


def get_oauth_settings():
    from slack_bolt.oauth.oauth_settings import OAuthSettings

    return OAuthSettings(
        client_id=os.environ["SLACK_CLIENT_ID"],
        client_secret=os.environ["SLACK_CLIENT_SECRET"],
        scopes=[
            "app_mentions:read",
            "channels:history",
            "channels:read",
            "chat:write",
            "chat:write.customize",
            "chat:write.public",
            "commands",
            "groups:read",
            "im:read",
            "users.profile:read",
            "users:read",
            "incoming-webhook",
            "mpim:read",
        ],
        install_page_rendering_enabled=False,
        installation_store=MongoDBInstallationStore(),
    )


@stub.function(image=image, secrets=secrets, keep_warm=1)
@modal.asgi_app(label=PREFIX + "-slack-bot")
def _asgi_app():
    from slack_bolt import App
    from slack_bolt.adapter.fastapi import SlackRequestHandler

    slack_app = App(oauth_settings=get_oauth_settings())

    fastapi_app = FastAPI()
    handler = SlackRequestHandler(slack_app)

    @slack_app.event("url_verification")
    def handle_url_verification(body, logger):
        challenge = body.get("challenge")
        return {"challenge": challenge}

    @slack_app.event("app_mention")
    def handle_app_mentions(body, say, client):
        print("here!")

    @slack_app.event("message")
    def handle_message(body, ack, message, client):
        ack()
        print(message)
        if "thread_ts" in message:
            # checking if the message is in a thread
            conversation = client.conversations_replies(
                channel=message["channel"], ts=message["thread_ts"]
            )
            thread_messages = conversation["messages"]
            if thread_messages:
                # checking if the thread has non-zero messages
                bot_profile = thread_messages[0].get("bot_profile")
                if thread_messages[-1].get("bot_profile"):
                    # prevent bots from replying to themselves
                    return
                if bot_profile and bot_profile.get("name") == "Sweep":
                    # checking that the message is from Sweep
                    channel_name = client.conversations_info(
                        channel=message["channel"]
                    )["channel"]["name"]
                    reply_slack.call(
                        SlackSlashCommandRequest(
                            channel_id=message["channel"],
                            channel_name=channel_name,
                            text=message["text"],
                            user_name=message["user"],
                            user_id=message["user"],
                            team_id=message["team"],
                            is_enterprise_install=False,
                        ),
                        thread_ts=message["thread_ts"],
                    )

    @slack_app.command("/sweep")
    def sweep(ack, respond, command, client):
        print(ack, respond, command, client)
        ack()
        respond("I'm working on it!")
        reply_slack.spawn(SlackSlashCommandRequest(**command))

    @fastapi_app.post("/")
    async def root(request: Request):
        return await handler.handle(request)

    @fastapi_app.get("/slack/install")
    async def oauth_start(request: Request):
        return await handler.handle(request)

    @fastapi_app.get("/slack/oauth_redirect")
    async def oauth_callback(request: Request):
        return await handler.handle(request)

    return fastapi_app
